File: backend/app/schemas/challenge.py
from pydantic import BaseModel


class ChallengeOption(BaseModel):
    text: str
    index: int
    is_correct: bool = False


class ChallengeQuestion(BaseModel):
    word: str
    target_word: str = ""  # 底层生字，前端原样回传给 /submit（level 2/4 时 word 为空）
    pinyin: str
    level: int
    level_name: str
    question_text: str
    hint: str
    options: list[ChallengeOption]
    correct_answer: str


class SubmitAnswerRequest(BaseModel):
    word: str
    subject: str
    grade: int = 3
    level: int
    answer: str
    streak: int = 0


class SubmitAnswerResponse(BaseModel):
    correct: bool
    correct_answer: str
    new_level: int
    points_earned: int
    streak: int
    badge_earned: str | None = None
    encouragement: str


class DailyTaskResponse(BaseModel):
    task_date: str
    subject: str
    total: int
    completed: int
    remaining: int
    words: list[str]


class BadgeDetail(BaseModel):
    id: str
    name: str
    desc: str


class UserStatsResponse(BaseModel):
    total_points: int
    streak_days: int
    words_mastered: int
    badges: list[BadgeDetail]

File: frontend/src/api/challenge.ts
import { request } from './config'

export interface ChallengeOption {
  text: string
  index: number
}

export interface ChallengeQuestion {
  word: string
  target_word: string
  pinyin: string
  level: number
  level_name: string
  question_text: string
  hint: string
  options: ChallengeOption[]
  correct_answer: string
}

export interface SubmitAnswerResponse {
  correct: boolean
  correct_answer: string
  new_level: number
  points_earned: number
  streak: number
  badge_earned: string | null
  encouragement: string
}

export interface DailyTaskResponse {
  task_date: string
  subject: string
  total: number
  completed: number
  remaining: number
  words: string[]
}

export interface BadgeDetail {
  id: string
  name: string
  desc: string
}

export interface UserStatsResponse {
  total_points: number
  streak_days: number
  words_mastered: number
  badges: BadgeDetail[]
}

export function getDailyTask(subject: string, grade: number): Promise<DailyTaskResponse> {
  return request<DailyTaskResponse>({
    url: `/challenge/daily?subject=${subject}&grade=${grade}`,
    method: 'GET',
  })
}

export function getChallengeQuestion(subject: string, grade: number): Promise<ChallengeQuestion | null> {
  return request<ChallengeQuestion | null>({
    url: `/challenge/question?subject=${subject}&grade=${grade}`,
    method: 'GET',
  })
}

export function submitChallengeAnswer(data: {
  word: string
  subject: string
  grade: number
  level: number
  answer: string
  streak: number
}): Promise<SubmitAnswerResponse> {
  return request<SubmitAnswerResponse>({
    url: '/challenge/submit',
    method: 'POST',
    data,
  })
}

export function getUserStats(): Promise<UserStatsResponse> {
  return request<UserStatsResponse>({ url: '/challenge/stats', method: 'GET' })
}
